### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by junekiwi** [maxmouchet/HMMBase.jl#2]
**March 11: The results below are outdated/wrong, please see [benchmark/](https://github.com/maxmouchet/HMMBase.jl/tree/master/benchmark) instead.**

Performance @ 3856e94. Tested on 10k observations.

```python
model = hmm.GaussianHMM(n_components=3, covariance_type="full")
model.startprob_ = np.array([.6, .3, .1])
model.transmat_ = np.array([[.7, .2, .1], [.3, .5, .2], [.3, .3, .4]])
model.means_ = np.array([[0.0, 0.0], [3.0, -3.0], [5.0, 10.0]])
model.covars_ = np.tile(np.identity(2), (3, 1, 1))
```

```julia
π0 = [.6, .3, .1]
P = [.7 .2 .1; .3 .5 .2; .3 .3 .4]
D = [MvNormal([0.0, 0.0], [1.0, 1.0]), MvNormal([3.0, -3.0], [1.0, 1.0]), MvNormal([5.0, 10.0], [1.0, 1.0])]
hmm = HMM(π0, P, D)
```

Times in ms. Mean time for Python via `%timeit`, median for Julia via `@benchmark`.

/ | hmmlearn | pyhsmm | HMMBase
--|--------------|-------------|----------------
Sample 10k | 981 | 189 | 1.818
lls | 1.11 |  | 3.562
forward (w/o lls computation) | 5.08 | | 1.510
backward (w/o lls computation) | 5.02 | | 1.402
Viterbi (w/o lls computation) | 0.319 | | 38.683
Viterbi | 1.99 | 78.2 | 45.182

**Comment by junekiwi** [comment]
Viterbi should be much faster in 0b993f3.

/ | hmmlearn | pyhsmm | HMMBase
--|--------------|-------------|----------------
Sample 10k | 981 | 189 | 1.818
lls | 1.11 |  | 3.562
forward (w/o lls computation) | 5.08 | | 1.510
backward (w/o lls computation) | 5.02 | | 1.402
Viterbi (w/o lls computation) | 0.319 | | **1.099**
Viterbi | 1.99 | 78.2 | **5.904**

**Comment by kmsquire** [comment]
Hello, the benchmark code in the repository only seems to show the forward and backward passes.  
Would you be able to share or check-in the code you're using to generate the tables above?

Also, I'm wondering what `lls` stands for...?

**Comment by junekiwi** [comment]
Hi,

You're right. I've added a few benchmarks (in [`benchmark/`](https://github.com/maxmouchet/HMMBase.jl/tree/master/benchmark)):
- `rand`
- `log_likelihoods`
- `messages_forwards`
- `messages_backwards`
- `viterbi`

Note that the tables above were generated manually and I think I've mixed some time units. The code/notebooks in `benchmark` should be more correct.  

`lls` is my shortcut for log-likelihoods. Since it is required for most of the algorithms, I separate the time it takes to compute the likelihood of each sample, vs. the time it takes to actually perform the core computations.
